### PR TITLE
Improve UX for logs to include SSH name and rank

### PR DESCRIPTION
**Pull request by iojw**
As discussed in #1369, this updates the debug logs to make it easier to figure out which node is the head node, which host name to use to ssh into a specific node, and also the rank of the node. With this PR, the debug output becomes the following:

```
(node-2, head, rank=0, pid=23862) 0
(node-2, head, rank=0, pid=23862) 172.31.95.105 172.31.81.91 172.31.85.131 172.31.89.126
(node-0, worker1, rank=1, pid=22755, ip=172.31.81.91) 1
(node-0, worker1, rank=1, pid=22755, ip=172.31.81.91) 172.31.95.105 172.31.81.91 172.31.85.131 172.31.89.126
(node-1, worker3, rank=3, pid=22760, ip=172.31.89.126) 3
(node-1, worker3, rank=3, pid=22760, ip=172.31.89.126) 172.31.95.105 172.31.81.91 172.31.85.131 172.31.89.126
(node-3, worker2, rank=2, pid=22757, ip=172.31.85.131) 2
(node-3, worker2, rank=2, pid=22757, ip=172.31.85.131) 172.31.95.105 172.31.81.91 172.31.85.131 172.31.89.126
```

with the following config:
```
resources:
  cloud: aws
  disk_size: 1024
  instance_type: m6i.large

num_nodes: 4

run: |
  echo $SKY_NODE_RANK
  echo $SKY_NODE_IPS
```

Thus, if the cluster is named `mycluster`, the user can easily figure out which host name to use to ssh into a node using the logs e.g. `ssh mycluster-worker1` for the log corresponding to `worker1`.

To ensure stableness for the naming of workers, we first sort them by their external IP, then do a mapping from internal to external IP in the code generation step to determine the node name. This means that the node name is always the same for a cluster regardless of what value is passed via the `num_nodes` argument. 

This also updates the logic for determining the rank of node to use the external IP of a node rather than the internal IP. However, as discussed in #1291, the rank is not always constant for a cluster, and is assigned on a per-task basis, as can be seen from the following output with `num_nodes` set to 2:

```
(node-0, worker3, rank=1, pid=24355, ip=172.31.89.126) 1
(node-0, worker3, rank=1, pid=24355, ip=172.31.89.126) 172.31.95.105 172.31.89.126
(node-1, head, rank=0, pid=26036) 0
(node-1, head, rank=0, pid=26036) 172.31.95.105 172.31.89.126
```
## Questions
- Do we still want to respect the `task_name` variable that is passed in and display it? The above output includes the task name.
- Open to suggestions on how to further improve the UX!

## Tested
- [x] Multi-node cluster
- [x] Single-node cluster

### CODE REVIEW

**Comment by concretevitamin** [comment]
This is awesome @iojw! Quick comments:

1. Maybe it's okay to get ride of `node-<i>` in the prefix now? It doesn't seem to help.
2. RE: "Do we still want to respect the task_name variable that is passed in and display it?" I feel it's okay to not display it in the logging prefix at all, for both multi- and single-node tasks.  For the latter, it doesn't seem helpful to display `head` as well.

**Comment by iojw** [comment]
@concretevitamin  I think one argument for keeping it in is that we currently use it to name the log path for multi-node clusters, and it is possible that users would want to figure out which log file contains the full logs for a node given the debug output. I don't think it would be possible to determine the rank from outside the codegen class to use as the logging path.

https://github.com/skypilot-org/skypilot/blob/a613b43a9f60848b8202486c3d52f4c147308fbc/sky/backends/cloud_vm_ray_backend.py#L3072

**Comment by concretevitamin** [comment]
Since `log_path` is eventually used inside the codegen

https://github.com/skypilot-org/skypilot/blob/3d3da862f299fd069d68c590cab3854f58740cc4/sky/backends/cloud_vm_ray_backend.py#L402

can we name the path using IP there?

**Comment by iojw** [comment]
@concretevitamin Ah yup! In that case I don't see any need for the notion of task name at all. I've updated the code to remove the notion of task name from the logs. The log files are now named according to the node name e.g. `head.log`, `worker1.log` etc.

The logs now look like so:
```
(worker1, rank=1, pid=23368, ip=172.31.87.209) 1
(worker1, rank=1, pid=23368, ip=172.31.87.209) 172.31.81.189 172.31.87.209 172.31.86.128 172.31.87.196
(worker3, rank=3, pid=23414, ip=172.31.87.196) 3
(worker3, rank=3, pid=23414, ip=172.31.87.196) 172.31.81.189 172.31.87.209 172.31.86.128 172.31.87.196
(head, rank=0, pid=26060) 0
(head, rank=0, pid=26060) 172.31.81.189 172.31.87.209 172.31.86.128 172.31.87.196
(worker2, rank=2, pid=23348, ip=172.31.86.128) 2
(worker2, rank=2, pid=23348, ip=172.31.86.128) 172.31.81.189 172.31.87.209 172.31.86.128 172.31.87.196
```

**Comment by iojw** [comment]
For single node tasks, the logs display the name of the task, or just `task` if no name is set.

```
(task, pid=32503) 0
(task, pid=32503) 172.31.95.145
```

**Comment by concretevitamin** [comment]
Thanks @iojw - one thought: how about we don't show the task name even for single-node tasks? This is to keep it consistent. Users can see their custom task names in `sky queue`.

**Comment by iojw** [comment]
@concretevitamin Ray does not play well with the `name` option being empty - if it is, the name in the logs defaults to the name of the remote function `run_bash_command_with_logs` instead. We can work around with passing in a space `" "` as the name, but this will lead to 2 spaces at the start of each log which is not ideal. Additionally, something to consider - it is not always the case that the task is executed on the head node for single-node tasks on multi-node clusters, so it may still make sense to include the node name. e.g. A single-node task may still be executed on `worker1` or `worker2`.

What do you think?

File: sky/backends/backend_utils.py
Comment (concretevitamin) [inline sky/backends/backend_utils.py:486]: This means `worker<i>` is sorted by public IPs, right? (Also, we should add "public" to the var name to be clearer.)

File: sky/backends/cloud_vm_ray_backend.py
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3056]: nit: internal_external_ips?

File: sky/backends/cloud_vm_ray_backend.py
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3062]: nit: This name could suggest this list itself is sorted. How about `stable_cluster_internal_ips`? Same for the arg on L3072.

File: sky/backends/cloud_vm_ray_backend.py
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3049]: These two calls may take some time IIRC. Do we want to cache L3060 somewhere?

**Comment by concretevitamin** [comment]
These are great points @iojw. How about for single-node tasks we just show `head/worker<i>` depending on which node is scheduled (no `task.name`)? This way we're consistent in both cases.

File: sky/backends/cloud_vm_ray_backend.py
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:334]: nit: remnant

File: sky/backends/cloud_vm_ray_backend.py
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:287]: It seems like we should pass this in for `_execute_task_one_node()` as well, in which case we can remove Optional for this arg.

File: sky/backends/cloud_vm_ray_backend.py
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:389]: 1. How about `<node_name>-<private ip>.log`? Just to give users more info if they need to debug.
2. I think L392 should use the same format, because of the reason you mentioned (we may have 1-node tasks in an n-node cluster).

**Review by concretevitamin — COMMENTED** [review]
Thanks @iojw! Did a pass.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:3062]: agreed!

File: sky/backends/backend_utils.py
Comment (iojw) [reply sky/backends/backend_utils.py:486]: Yup, we sort based on public IPs.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:287]: Agreed! That would make reasoning about this a lot easier.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:3049]: One way would be to cache it in the `ResourceHandle` but I'm not sure if / how we would decide to invalidate the cache. Wouldn't it be possible that one of the workers dies and gets restarted? Is there a callback or path somewhere which would allow us to handle this case?

**Comment by iojw** [comment]
Thanks @concretevitamin! To summarize the changes, there are no longer any differences between single-node and multi-node **tasks** on multi-node **clusters** in the logging output and log path. The difference lies between single-node **clusters** and multi-node **clusters**.

For single-node clusters:
- For logging, we display the task name, or just `task` if the name is not set. (no rank)
- We store logs at `run.log`
- This is so that users who primarily use single-node clusters need not know the concept of head nodes at all

For multi-node clusters:
- For logging, we always display node name along with rank, regardless of `num_nodes`
- We store logs at `node_name-ip.log`, where IP is the private ip of the node

I've also updated `ResourceHandle` so that we able to cache the value of the `stable_cluster_internal_ips` variable after it is retrieved the first time. Unsure if there might be an issue when workers get restarted since we do not invalidate the cache.

SIde-note: I think this also makes the code is a lot clearer!

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py]: Pushed a commit to handle this 👍.  We do need to invalidate the entry where appropriate.

File: sky/backends/backend_utils.py
Comment (Michaelvll) [inline sky/backends/backend_utils.py:1158]: Do we want to cache both internal and external IPs, so that in the `cloud_vm_ray_backend`, we can use those ips to run `setup` or `execute`?

File: sky/backends/backend_utils.py
Comment (Michaelvll) [inline sky/backends/backend_utils.py:1670]: `get_node_ips` can be quite time consuming, can we reuse the `external_ips` we get from L1648?

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:441]: It may be better to have the rank in the name or have both `ip` and rank, instead of only having the `ip`, because the user can infer infer the IP from the `node_name`, but may not be easy to get the rank information. How about `{{rank}}-{{node_name}}.log`?

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:1549]: As mentioned above, can we cache both `external` and `internal` ips?

It can be used to speed up the places we call `get_node_ips` in the `cloud_vm_ray_backend`. With that said, I would prefer to add another argument called `use_cached_ips: bool` to the `backend_utils.get_node_ips`, so we can extend the following lines to use the cached ips.
https://github.com/skypilot-org/skypilot/blob/b2d3555af1f4b5aba53cc0c04c39ccd44fa26bc1/sky/backends/backend_utils.py#L1152-L1155

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:1560]: This head_ip can be stale. How about we have a `@property` function called `head_ip` instead
```
@property
def head_ip(self):
    return self.stable_ips[0][1]
```

This needs to be careful due to the backward compatibility, for loading existing handles from the our cluster database.

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick fix @iojw! These changes can potentially reduce the overhead of our system for launching tasks on multi-node clusters. Left several comments.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1560]: When would this `heap_ip` value be stale? From what I see, it's updated and invalidated at the same places as the stable ip list currently.

Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1560]: Oh, I was trying to say that if we are going to cache the external IP as well, the `head_ip` is a duplicated variable that may not be needed.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1549]: Since the `get_stable_cluster_ips()` function now calls `get_node_ips()`, what do you think about adding the `use_cached_ips` argument to the `get_stable_cluster_ips()` function and replacing all `get_node_ips()` calls with `get_stable_cluster_ips()` calls instead? This is so that `get_stable_cluster_ips()` is now our "top-level" function. I think this makes more sense because otherwise we would have both functions calling each other in different cases.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1560]: That makes sense!

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1549]: Good point! I think it should be fine to replace the `get_node_ips` with `get_stable_cluster_ips`.

**Comment by iojw** [comment]
I've updated the code to cache a stable list of (internal, external) IP tuples instead, as well as changed the `head_ip` attribute of `ResourceHandle` to be an `@property` instead. This turns out to be a non-trivial refactor so please take a look!

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:1874]: In many other places, we will check if the `head_ip` is set in the `handle` to determine whether the cluster is up, e.g. https://github.com/skypilot-org/skypilot/blob/6a48a127e73dc0cdf452c2db3e8eca2de19c0fbb/sky/core.py#L340

and https://github.com/skypilot-org/skypilot/blob/6a48a127e73dc0cdf452c2db3e8eca2de19c0fbb/sky/core.py#L611-L612

We need to cache the head_ip into the handle after provisioning.

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:1567]: I would prefer to have this `stable_internal_external_ips` private, and only expose the `interal_ips` and `external_ips` as properties of this class, to make the code more readable and easier to maintain.

**Review by Michaelvll — COMMENTED** [review]
Thank you for the refactoring and the great effort to make this work @iojw! This can improve the user experience a lot for the mutli-node users. Left several comments. Some are for correctness but most of them are small coding style nits.

File: sky/backends/backend_utils.py
Comment (iojw) [reply sky/backends/backend_utils.py:1686]: You're right, thank you for catching this

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1874]: Doesn't this mean that we cannot remove `head_ip` and replace it with as an `@property` variable?

File: sky/backends/backend_utils.py
Comment (iojw) [reply sky/backends/backend_utils.py:1684]: I see - will update the call on L1666 to not use cached ips instead

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1565]: To allow for retrieving the stable list inside `__setstate__`, I think another way would be to call the `get_stable_cluster_ips` function on the handle itself after its attributes have all been updated - let me know what you think

File: sky/backends/backend_utils.py
Comment (iojw) [reply sky/backends/backend_utils.py:1270]: Ah I was not aware of that - reverted the change

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1331]: It seems that there are optimizations in `_gang_schedule_ray_up` that rely on the`num_nodes` variable e.g.

https://github.com/skypilot-org/skypilot/blob/396eb0967f27634f2d2c2b531f380b67d4933709/sky/backends/cloud_vm_ray_backend.py#L1334

`wait_until_ray_cluster_ready`, which `_gang_schedule_ray_up` calls also relies on this variable

https://github.com/skypilot-org/skypilot/blob/396eb0967f27634f2d2c2b531f380b67d4933709/sky/backends/backend_utils.py#L898

Could you elaborate? To me, it does not seem like the argument can be removed cleanly

**Comment by concretevitamin** [comment]
Is this good to go?  @Michaelvll 

There's some conflicts to resolve.

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1331]: I think the `num_nodes` can be got from `cluster_handle.launched_nodes`?

Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1331]: Ahh I see, got it thank you!

**Comment by iojw** [comment]
Thanks @Michaelvll! I believe there's just one remaining comment regarding how we should handle `head_ip` in `ResourceHandle`.

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1910]: Could you explain more?

Previously, in L1877, we have the `head_ip`for creating the `handle`, but now we removed it. I mean why cannot us have the `ip_list` cached in the handle after it is provisioned?

Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py]: I see, I didn't realize you were referring to the specific case of a single-node cluster. I believe we cannot use the optimization here anymore because we need both internal and external IP of the head node for single-node clusters, but parsing the Ray output can only give us the external IP. I've updated the code so that `get_stable_cluster_ips` is always called after provisioning, and this will cache the ip list in the handle.

File: sky/backends/backend_utils.py
Comment (Michaelvll) [inline sky/backends/backend_utils.py:1200]: The L1212-L1214 seems not needed now, since the caching is mechanism is already handled by the `get_stable_cluster_ips()`

**Review by Michaelvll — COMMENTED** [review]
Thank you for the quick fix @iojw! It looks great! I am testing it now. I feel like we need to refactor the `get_stable_cluster_ips` into the `ResourcesHandle` class to make the code more readable. Please refer to the comment below. I think once the comments are addressed and the test succeeded, it should be ready to go. : )

File: sky/backends/backend_utils.py
Comment (iojw) [reply sky/backends/backend_utils.py:1143]: Yup, you are right - these were left over from a previous commit.

File: sky/backends/backend_utils.py
Comment (iojw) [reply sky/backends/backend_utils.py:1146]: I believe there are places where we want to pass in different values e.g. L1908 and L2077 in `cloud_vm_ray_backend.py` where we use `_HEAD_IP_MAX_ATTEMPTS` and `_WORKER_IP_MAX_ATTEMPTS`. Currently they are the same value but we may potentially want to pass in different values in the future?

Comment (Michaelvll) [reply sky/backends/backend_utils.py:1146]: It should be fine to keep them always the same, as we don't need to handle them differently for the head and workers. Maybe we can worry about the separation of the two values for the attempts when we actually found separating them is important in the future. : )

File: sky/backends/backend_utils.py
Comment (iojw) [reply sky/backends/backend_utils.py]: That makes sense, thanks!

**Review by Michaelvll — APPROVED** [review]
Awesome! Thanks for the quick fix @iojw! Left several final comments. Please run the smoke test before merging this PR, since it can affect all the cases.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1680]: Yes, the only requirement we have here is that the order of internal IPs matches that of external IPs. 

Looking at https://github.com/ray-project/ray/blob/30a09bd66c46f29e1463287982bd03afc7ecda19/python/ray/autoscaler/_private/commands.py#L1305, this is the case

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1931]: You are right, thank you for catching this!

**Comment by iojw** [comment]
@Michaelvll I can run the smoke tests for AWS and Azure but not GCP since I don't have GCP credentials - could you help me run them for GCP?

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1679]: It seems to depend on the order the following function returns each time it is called:
https://github.com/ray-project/ray/blob/30a09bd66c46f29e1463287982bd03afc7ecda19/python/ray/autoscaler/_private/commands.py#L1315
And the above function depends on the `ec2` function for filtering the nodes by the tag:
https://github.com/skypilot-org/skypilot/blob/235d9d65f5e269a10cbd7bec848a89d830d1d8a0/sky/skylet/providers/aws/node_provider.py#L157-L158
Could we find the doc for this function to make sure the order is deterministic for multiple call?

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:1732]:
```suggestion
            if version < 3:
                self.stable_internal_external_ips = None
                if head_ip is not None:
                    self._update_stable_cluster_ips()
```

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:1726]:
```suggestion
                head_ip = state.pop('head_ip', None)
                self.stable_internal_external_ips = None
```

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1679]: I see - I could not find any guarantee that this function is deterministic for multiple calls. There is also the case of the providers for GCP and Azure, which both call their own respective libraries. I tried looking around and I didn't see a good way of doing this without using Ray's private APIs. What do you think?

Perhaps we could submit a PR or patch `get_worker_node_ips` so that the nodes are sorted?

**Comment by Michaelvll** [comment]
We got several smoke tests issues for incorrect cluster status refresh and some internal/external IP length mismatch problem:
https://gist.github.com/Michaelvll/c7f3891653ded20f0aa3cd41562bc285
https://gist.github.com/Michaelvll/1bfeeb379c355f7fde3baeab1a04f90d
Let's hold this PR from merging until we fix the test errors. : )

**Comment by Michaelvll** [comment]
Thanks for the quick fix @iojw! Just tried the `sky status` that has an existing cluster, and it worked. I am running the smoke test for this and the `tests/backward_comaptibility_tests.sh`.

Tested:
- [ ] `tests/run_smoke_tests.sh`
- [ ] `tests/backward_comaptibility_tests.sh`

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1695]: Yea, that could be future work. How about we try to run it multiple times and see if the external_ips and internal_ips always match each other? If that works, probably it is fine for now.

File: sky/backends/cloud_vm_ray_backend.py
Comment (iojw) [reply sky/backends/cloud_vm_ray_backend.py:1687]: Nice!

**Review by Michaelvll — APPROVED** [review]
Thank you for the awesome work @iojw! It passed all the tests on my side. I think it is good to go.

**Comment by iojw** [comment]
Thank you @Michaelvll @concretevitamin for the reviews!

**Pull request merged**
